matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessagesAdapter.java

```diff
@@ -851,25 +851,60 @@ public MessageRow getClosestRow(Event event) {
      */
     public MessageRow getClosestRowFromTs(final String eventId, final long eventTs) {
         MessageRow messageRow = getMessageRow(eventId);
-        if (messageRow != null) {
-            return messageRow;
-        } else {
-            ArrayList<MessageRow> rows = new ArrayList<>(mEventRowMap.values());
-            for (MessageRow row : rows) {
-                if (row.getEvent().getOriginServerTs() > eventTs) {
-                    if (messageRow == null) {
-                        messageRow = row;
-                    } else if (row.getEvent().getOriginServerTs() - eventTs <
-                            messageRow.getEvent().getOriginServerTs() - eventTs) {
-                        messageRow = row;
-                        Log.d(LOG_TAG, "getClosestRow " + row.getEvent().eventId);
-                    }
+
+        if (messageRow == null) {
+            messageRow = getClosestRow(new ArrayList<>(mEventRowMap.values()), eventTs);
+        }
+
+        return messageRow;
+    }
+
+    /**
+     * Get the closest row before the given event id/ts
+     *
+     * @param eventId
+     * @param eventTs
+     * @return closest row
+     */
+    public MessageRow getClosestRowBeforeTs(final String eventId, final long eventTs) {
+        MessageRow messageRow = getMessageRow(eventId);
+
+        if (messageRow == null) {
+            List<MessageRow> rows = new ArrayList<>(mEventRowMap.values());
+            Collections.reverse(rows);
```

Comment:
- The reverse is useless since the list is made from a map valuse so the order is random
- Instead, the conditions should be updated in getClosestRow